app/views/service_worker/index.js.erb

```diff
@@ -1,144 +1,132 @@
         }
-        garbageCollect(cache);
-      })
-    return response;
-  }
+        if (!response.ok && response.status != 404) {
+          return caches.match('/500.html');
+        }
+        return response;
+      }).catch(err => caches.match('/offline.html'));
 
-  function garbageCollect(cache) {
-    cache.keys().then(function count(keys) {
-      if (keys.length > REQUESTS_LIMIT) {
-        cache.delete(keys[0]);
-        cache.delete(keys[1]);
-        cache.delete(keys[2]);
-        cache.delete(keys[3]);
-        cache.delete(keys[4]);
-        cache.delete(keys[5]);
-        cache.delete(keys[6]);
-        cache.delete(keys[7]);
-        cache.delete(keys[8]);
-        cache.delete(keys[9]);
-        cache.delete(keys[10]);
+      function pushStream(stream) {
+        const reader = stream.getReader();
+        return reader.read().then(function process(result) {
+          if (result.done) return;
+          controller.enqueue(result.value);
+          return reader.read().then(process);
+        });
       }
-    })
-  }
+      startFetch
+        .then(response => pushStream(response.body))
+        .then(() => middleFetch)
+        .then(response => pushStream(response.body))
+        .then(() => endFetch)
+        .then(response => pushStream(response.body))
+        .then(() => controller.close());
+    }
+  });
 
-  function unableToResolve () {
-    console.log('WORKER: fetch request failed in both cache and network.');
-    return caches.match('/offline.html')
-  }
-};
+  return new Response(stream, {
+    headers: {'Content-Type': 'text/html; charset=utf-8'}
+  });
+}
 
-function onPush(event) {
-  var title = 'DEV Connect 👋 ';
-  var body = event.data.text();
-  var icon = '<%= image_path "devlogo-pwa-128.png" %>';
-  var tag = 'default-tag' + body;
-  event.waitUntil(
-    self.registration.showNotification(title, {
-      body: body,
-      icon: icon,
-      tag: tag,
-      data: {
-        url: 'https://dev.to/connect'
+self.addEventListener('fetch', event => {
+  const url = new URL(event.request.url);
+  if (url.origin === location.origin) {
+    // 
+    if (event.clientId === "" && // Not fetched via AJAX after page load.
+      event.request.method == "GET" && // Don't fetch on POST, DELETE, etc.
+      !event.request.referrer.includes('/signout_confirm') && // If this is the referrer, we instead want to flush.
+      !url.href.includes('i=i') && // Parameter representing "internal" navigation.
+      !url.href.includes('?signin') && // Don't run on sign in.
+      !url.href.includes('/onboarding') && // Don't run on onboarding.
+      !url.href.includes('/api/') && // Don't run on API endpoints.
+      !url.href.includes('/users/auth') && // Don't run on authentication.
+      !url.href.includes('/shell_') && // Don't fetch for shell.
+      !url.href.includes('/admin') && // Don't fetch for administrate dashboard.
+      !url.href.includes('/internal') && // Don't fetch for internal dashboard.
```

Comment:
We probably could. I feel like either way you're removing colocation which could come in handy in one place or another. Making it explicit here is leaving it up to the serviceworker developers to make sure their code plays well with the server code, rather than the other way around. This seems somewhat correct to me.

These are all pretty special cases here. We're not going to be adding new controllers that could break this very often.